### PR TITLE
Make builder scheduler target-aware

### PR DESCRIPTION
**Pull request by chefsalim**
This change adds target (ie, platform+architecture) awareness to the builder scheduler. Now there are separate dependency graphs per target, instead of having all the package dependencies in a single graph. 

Note - a couple of builder REST APIs will need to eventually be changed to incorporate the target as a parameter. For now, these default to x86_64-linux which is what the builder currently supports (we build a graph for Windows packages for future builds).

Signed-off-by: Salim Alam <salam@chef.io>

tenor-114178714

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-core/src/target_graph.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+pub struct TargetGraph {
+    graphs: HashMap<PackageTarget, PackageGraph>,
+}
+
+impl TargetGraph {
+    pub fn new() -> Self {
+        let mut graphs = HashMap::new();
+
+        // We only support the following targets currently
+        for target_str in vec!["x86_64-linux", "x86_64-windows"] {
```
Comment (georgemarshall) [inline components/builder-core/src/target_graph.rs]: An array or tuple might be better served here

File: components/builder-scheduler/src/data_store.rs [hunk 2]
```diff
@@ -391,6 +391,31 @@ impl DataStore {
...
+            r#"CREATE OR REPLACE FUNCTION upsert_package_v1 (
+                                    in_ident text,
+                                    in_deps text[],
+                                    in_target text
+                                ) RETURNS SETOF packages AS $$
+                                        BEGIN
+                                            RETURN QUERY INSERT INTO packages (ident, deps, target)
+                                            VALUES (in_ident, in_deps, in_target)
+                                            ON CONFLICT(ident)
+                                            DO UPDATE SET deps=in_deps, target=in_target RETURNING *;
```
Comment (raskchanky) [inline components/builder-scheduler/src/data_store.rs]: Nice!

Comment (chefsalim) [reply components/builder-scheduler/src/data_store.rs]: Yeah it's pretty cool.. this capability was added in Postgres 9.5

**Comment by chefsalim** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/273386781).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/273386781).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/273386781).

**Comment by thesentinels** [comment]
:broken_heart: Travis CI reports [this PR failed to pass the test suite](https://travis-ci.org/habitat-sh/habitat/builds/273386781).

The next step is to examine the [job](https://travis-ci.org/habitat-sh/habitat/builds/273386781) and figure out why. If it is transient, you can try re-triggering the Travis CI Job - if it passes, this PR will be automatically merged. If it is not transient, you should fix the issue and update this pull request, and issue `approve` again. If you believe it will never pass, and you are feeling :godmode:, you can issue a `force` to merge this PR anyway.

**Pull request merged**